File: 445.两数相加-ii.cpp
/*
 * @lc app=leetcode.cn id=445 lang=cpp
 *
 * [445] 两数相加 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
  public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
        l1 = reverseList(l1);
        l2 = reverseList(l2);

        int f = 0;
        auto result = new ListNode(0);
        auto p = result;
        while (l1 && l2) {
            int n = l1->val + l2->val + f;
            f = n >= 10 ? 1 : 0;
            p->next = new ListNode(f ? n % 10 : n);
            p = p->next;
            l1 = l1->next;
            l2 = l2->next;
        }
        if (f) {
            if (l1) {
                while (l1 && f) {
                    int n = l1->val + f;
                    f = n >= 10 ? 1 : 0;
                    p->next = new ListNode(f ? n % 10 : n);
                    p = p->next;
                    l1 = l1->next;
                }
                if (l1) {
                    p->next = l1;
                }
            }
            if (l2) {
                while (l2 && f) {
                    int n = l2->val + f;
                    f = n >= 10 ? 1 : 0;
                    p->next = new ListNode(f ? n % 10 : n);
                    p = p->next;
                    l2 = l2->next;
                }
                if (l2) {
                    p->next = l2;
                }
            }

            if (f) {
                p->next = new ListNode(f);
            }
        } else {
            p->next = l1 ? l1 : l2;
        }

        return reverseList(result->next);
    }

    ListNode *reverseList(ListNode *head) {
        if (!head || !head->next)
            return head;
        ListNode *cur = head;
        ListNode *next = head->next;
        cur->next = NULL;
        ListNode *p = cur->next, *q = next->next;
        while (next) {
            cur->next = p;
            q = next->next;
            next->next = cur;
            p = cur;
            cur = next;
            next = q;
        }
        return cur;
    }
};
// @lc code=end

File: 445.两数相加-ii.go
/*
 * @lc app=leetcode.cn id=445 lang=golang
 *
 * [445] 两数相加 II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

 func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var v1, v2 []int
	var ret = &ListNode{}
	for l1 != nil {
		v1 = append(v1, l1.Val)
		l1 = l1.Next
	}
	for l2 != nil {
		v2 = append(v2, l2.Val)
		l2 = l2.Next
	}
	n1, n2, i, j := len(v1)-1, len(v2)-1, 0, 0
	carry := 0
	for i, j = n1, n2; i >= 0 && j >= 0; {
		r := v1[i] + v2[j] + carry
		if r > 9 {
			carry = 1
		} else {
			carry = 0
		}
		ret.Val = r%10
		ret = &ListNode{Next: ret}
		i--
		j--
	}
	for j >= 0 {
		r := v2[j] + carry
		if r > 9 {
			carry = 1
		} else {
			carry = 0
		}
		ret.Val = r%10
		ret = &ListNode{Next: ret}
		j--
	}
	for i >= 0 {
		r := v1[i] + carry
		if r > 9 {
			carry = 1
		} else {
			carry = 0
		}
		ret.Val = r%10
		ret = &ListNode{Next: ret}
		i--
	}
	if carry == 1 {
		ret.Val = carry % 10
		ret = &ListNode{Next: ret}
	}
	return ret.Next
}
// @lc code=end

